### PR TITLE
Fix: The IPv6 address detection should ignore unique local IPv6 Address

### PR DESCRIPTION
**Pull request by BSoD-Ultimate**
Some network configurations might let the gateway assigns Unique Local Addresses to devices connected to the local network. 
The purpose of assigning this kind of address is to help devices on the same local network could communicate more conveniently since the IPv6 address on the Internet might change over time. This kind of IPv6 address could not be accessed outside the local network. 

See RFC 4193: https://tools.ietf.org/html/rfc4193

### CODE REVIEW

**Comment by LINKIWI** [comment]
Thanks. @Mygod any thoughts?

File: cloudflare-ddns [hunk 1]
```diff
@@ -142,13 +142,34 @@ def get_ipv6():
...
+    # Some network configurations might allow the gateway to provide unique local addresses(ULA) to the devices on the LAN.
+    # In this case, one network adapter could have two non-temporary IPv6 addresses:
+    #   The IPv6 Address on the Internet(2000::/3) and The unique local address.
+    #   The second one cannot be accessed outside, it is designed to provide convenience on accessing devices on the same network.
+    # The ULA prefix is statically configured by the gateway, which means devices on the same local network could have
+    # a fixed IPv6 address locally even though the IPv6 address prefix on the Internet might change over time.
+    # According to rfc 4193, the IPv6 address range fc00::/7 is reserved for this kind of purpose.
+    def is_ula_addr(ipAddr):
+        # Only check the first 16 bits of the IPv6 address to judge if the address is one of the ULA addresses
+        return (int(ipAddr[:4], 16) & 0xfe00 ) == 0xfc00
```
Comment (Mygod) [inline cloudflare-ddns]: `int` could throw exceptions.

File: cloudflare-ddns [hunk 2]
```diff
@@ -142,13 +142,34 @@ def get_ipv6():
...
+
+    matchedIPs = []
+
     inet6_finder = re.compile('^    inet6 ([0-9a-f:]+)')
-    for line in subprocess.check_output(['ip', '-6', 'addr', 'list', 'scope', 'global', '-deprecated']).decode('utf-8').split('\n'):
+
+    # add the 'mngtmpaddr' argument to filter temporary IPv6 addresses
+    for line in subprocess.check_output(['ip', '-6', 'addr', 'list', 'scope', 'global', 'mngtmpaddr', '-deprecated']).decode('utf-8').split('\n'):
```
Comment (Mygod) [inline cloudflare-ddns]: `mngtmpaddr` should not be added. Using privacy extension IPv6 address is the intended default behavior.

File: cloudflare-ddns [hunk 3]
```diff
@@ -142,13 +142,34 @@ def get_ipv6():
...
         if match is not None:
             # Multiple address might be present, assuming the first one is the best
-            # Maybe add flag for preventing temporary addresses?
-            return match.group(1)
+            ipv6_addr = match.group(1)
+            if not is_ula_addr(ipv6_addr):
+                matchedIPs.append(ipv6_addr)
+
+    if len(matchedIPs) > 0:
+        return matchedIPs[0]
```
Comment (Mygod) [inline cloudflare-ddns]: Unnecessary change.

**Review by Mygod — CHANGES_REQUESTED** [review]

**Comment by BSoD-Ultimate** [comment]
Assuming the `ip addr` command will always give the program valid IPv6 addresses.

I have understood that selecting privacy extension IPv6 addresses is the intended behavior.

My comprehension of your suggestions is that I should return the function immediately after the first non-ULA IPv6 address is found. The function `is_ula_addr` needs to do more checks to get the first 16 bits of the address correctly. For example, IPv6 addresses in 0800::/5 might be used as global unicast addresses in the future, so just capturing the first 4 characters in the string is not a good way. Also, I must take care of exceptions if the checks in `is_ula_addr` throw them.

Is this right?

**Comment by Mygod** [comment]
Looks good.

**Comment by Mygod** [comment]
However, I think those addresses should already be filtered by global, no?

**Comment by BSoD-Ultimate** [comment]
No, these ULA addresses are not filtered. I use OpenWRT as the gateway in my room. The OpenWRT configuration and things that happened in my place are below.
OpenWRT configuration:
image
Linux:
image
Windows:
image
